The Gut Phage Database (GPD)
============================

Scripts used for characterizing human gut bacteriophages in the following manuscript:

Camarillo-Guerrero LF, Almeida A, Rangel-Pineros G, RD Finn, Lawley TD (2020) [Massive expansion of human gut bacteriophage diversity]

Associated data can also be found in our [FTP server](http://ftp.ebi.ac.uk/pub/databases/metagenomics/genome_sets/gut_phage_database/)

## Other analysis and plotting scripts

<b>scripts/</b>
* 'Figure 1.py': Distribution of MIUViG scores from CheckV analysis
* 'Figure 2.py': Viral diversity patterns across gut bacteria genera and broad host range VCs
* 'Figure 3.py': Gut phageome profiling across human populations and correlation with gut bacteria enterotypes
* 'Figure 4.py': Crass-like family global distribution and host-phage network of globally distributed VCs
* 'Figure 5.py': Phylogenetic structure of the pX phage and global distribution
* 'Figure S1.py': Quality control assessment of GPD
* 'Figure S2.py': Viral diversity patterns across gut bacteria phyla and host range analysis of gut phages
* 'Figure S3.py': Correlation between sequencing depth and number of phages detected in a sample
* 'Figure S4.py': Lachnospirales high connectivity 
